### PR TITLE
[Focus states] Use :focus-visible over :focus for focus-ring state.

### LINKED ISSUE #8053: Update Polaris components to only show focus state when navigating with keyboard
We currently use a focus-ring to show a merchant which element is currently in focus. This is useful particularly for users that don't use a traditional pointer device to navigate through the admin. We currently use the `:focus` CSS selector to add these focus rings to interactive elements. 

Whilst this is an important accessibility necessity, it can leave the admin looking very busy for merchants that do use a traditional pointer device, with focus rings appearing on every interaction. 

There is a CSS selector named `:focus-visible`, which allows the browser to determine when to show the focus state depending on the input device that controls the focus state, and the element in question. This will allow us to persist the vital focus state visibility for users navigating the admin via keyboard, but will leave the admin feeling cleaner for those merchants using a mouse or other pointer device.

https://github.com/Shopify/polaris/issues/8053#tasklist-block-7640fea2-eaff-49d9-bb62-5c59ffc9de0d

### PR DESCRIPTION
**Pull request by mrcthms**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris/issues/8053
Related to: https://github.com/Shopify/polaris/issues/7585 and https://github.com/Shopify/polaris/issues/818
We currently use a focus-ring to show a merchant which element is currently in focus. This is useful particularly for users that don't use a traditional pointer device to navigate through the admin. We currently use the :focus CSS selector to add these focus rings to interactive elements.

Whilst this is an important accessibility necessity, it can leave the admin looking very busy for merchants that do use a traditional pointer device, with focus rings appearing on every interaction.

There is a CSS selector named :focus-visible, which allows the browser to determine when to show the focus state depending on the input device that controls the focus state, and the element in question. This will allow us to persist the vital focus state visibility for users navigating the admin via keyboard, but will leave the admin feeling cleaner for those merchants using a mouse or other pointer device.

### WHAT is this pull request doing?

This PR updates instances of `:focus` to show a focus ring and replaces them with `:focus-visible`. It also tidies up some legacy code in some components to detect for keyFocus only, preferring to use the browser standard `:focus-visible` rather than some internal component logic to figure out when focus is being managed by a keyboard.

_Note: The Listbox remains untouched as the focus state logic there is pretty custom and will require somebody with more domain knowledge to update that component_

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: documentation/guides/legacy-polaris-v8-public-api.scss [hunk 1]
```diff
@@ -1347,7 +1347,7 @@ $nav-min-window: em(layout-width(page-with-nav));
     @include high-contrast-outline;
   }
 
-  &:focus {
+  &:focus-visible {
```
Comment (chloerice) [inline documentation/guides/legacy-polaris-v8-public-api.scss]: This code doesn't need to be changed as it isn't published, it's just a reference to the legacy Sass styles.

**Review by chloerice — APPROVED** [review]
Tophatted all affected components and all are as expected between keyboard and mouse focus, thanks for tackling this @mrcthms!! 🙌🏽 I'll update the Listbox, it should only show the surface-selected state and not use focus-ring (purely a CSS change)

**Comment by mrcthms** [comment]
Thanks @chloerice, much appreciated! LMK if you want a reviewer of the Listbox update.

**Pull request merged**